File: web/src/sim/retry.ts
// Port of conduit/core/retry.py: classification and exponential backoff with full jitter.

import type { RetryPolicy } from "./specs";
import type { Rng } from "./prng";

export class DeliveryError extends Error {
  readonly retryable: boolean = false;
  constructor(message: string, readonly status: number | null = null) {
    super(message);
    this.name = "DeliveryError";
  }
}

/** 429, 5xx, timeouts, connection resets: retry with backoff. */
export class TransientError extends DeliveryError {
  override readonly retryable = true;
  constructor(message: string, status: number | null = null, readonly retryAfter: number | null = null) {
    super(message, status);
    this.name = "TransientError";
  }
}

/** Hard 4xx (bad payload, auth, not found): do not retry, let SQS redrive. */
export class PermanentError extends DeliveryError {
  override readonly retryable = false;
  constructor(message: string, status: number | null = null) {
    super(message, status);
    this.name = "PermanentError";
  }
}

export interface FakeResponse {
  status: number;
  headers: Record<string, string>;
  body: Record<string, unknown>;
}

export function classifyResponse(response: FakeResponse, policy: RetryPolicy): DeliveryError | null {
  const status = response.status;
  if (status >= 200 && status < 300) return null;
  const body = JSON.stringify(response.body).slice(0, 200);
  if (policy.retryOnStatus.includes(status) || status >= 500) {
    return new TransientError(`HTTP ${status}: ${body}`, status, parseRetryAfter(response.headers["Retry-After"]));
  }
  return new PermanentError(`HTTP ${status}: ${body}`, status);
}

export function classifyException(exc: unknown): DeliveryError {
  if (exc instanceof DeliveryError) return exc;
  const name = exc instanceof Error ? exc.name : "Error";
  const message = exc instanceof Error ? exc.message : String(exc);
  if (/Timeout|Network|RemoteProtocol/.test(name)) return new TransientError(`${name}: ${message}`);
  return new PermanentError(`${name}: ${message}`);
}

function parseRetryAfter(value: string | undefined): number | null {
  if (!value) return null;
  const n = Number(value);
  return Number.isFinite(n) ? Math.max(0, n) : null;
}

/** Full-jitter delay for attempt (1-based): uniform(0, min(cap, base * mult^(n-1))). */
export function backoffCeiling(attempt: number, policy: RetryPolicy): number {
  if (attempt < 1) throw new Error("attempt is 1-based");
  return Math.min(policy.maxSeconds, policy.baseSeconds * policy.multiplier ** (attempt - 1));
}

export function backoffDelay(attempt: number, policy: RetryPolicy, rng: Rng): number {
  return rng.uniform(0, backoffCeiling(attempt, policy));
}

export interface RetryAttempt {
  attempt: number;
  outcome: "ok" | "transient" | "permanent" | "exhausted";
  status: number | null;
  error: string | null;
  /** Delay slept before the next attempt, when there is one. */
  delay: number | null;
  ceiling: number | null;
  retryAfter: number | null;
}

export interface RetryOutcome {
  attempts: number;
  delays: number[];
  trace: RetryAttempt[];
}

/**
 * Call fn until it succeeds, raises a PermanentError, or exhausts attempts.
 * Only TransientError triggers a retry; the sleep is virtual.
 */
export async function retryCall<T>(
  fn: () => Promise<T>,
  policy: RetryPolicy,
  rng: Rng,
  sleep: (seconds: number) => void,
  onRetry?: (attempt: number, delay: number, err: TransientError) => void,
): Promise<{ value: T; outcome: RetryOutcome }> {
  const outcome: RetryOutcome = { attempts: 0, delays: [], trace: [] };
  for (;;) {
    outcome.attempts += 1;
    try {
      const value = await fn();
      outcome.trace.push({ attempt: outcome.attempts, outcome: "ok", status: 200, error: null, delay: null, ceiling: null, retryAfter: null });
      return { value, outcome };
    } catch (exc) {
      const err = classifyException(exc);
      if (!err.retryable || outcome.attempts >= policy.maxAttempts) {
        outcome.trace.push({
          attempt: outcome.attempts,
          outcome: err.retryable ? "exhausted" : "permanent",
          status: err.status,
          error: err.message,
          delay: null,
          ceiling: null,
          retryAfter: null,
        });
        (err as DeliveryError & { outcome?: RetryOutcome }).outcome = outcome;
        throw err;
      }
      const transient = err as TransientError;
      const ceiling = backoffCeiling(outcome.attempts, policy);
      let delay = rng.uniform(0, ceiling);
      if (transient.retryAfter !== null) delay = Math.min(policy.maxSeconds, Math.max(delay, transient.retryAfter));
      outcome.delays.push(delay);
      outcome.trace.push({
        attempt: outcome.attempts,
        outcome: "transient",
        status: err.status,
        error: err.message,
        delay,
        ceiling,
        retryAfter: transient.retryAfter,
      });
      onRetry?.(outcome.attempts, delay, transient);
      sleep(delay);
    }
  }
}

File: conduit/core/retry.py
"""Retry classification and exponential backoff with full jitter."""

from __future__ import annotations

import random
import time
from collections.abc import Callable, Iterator
from dataclasses import dataclass, field
from typing import TypeVar

import httpx

from conduit.config import RetryPolicy

T = TypeVar("T")


class DeliveryError(Exception):
    """Base for adapter failures. ``retryable`` decides the worker's response."""

    retryable = False

    def __init__(self, message: str, *, status: int | None = None) -> None:
        super().__init__(message)
        self.status = status


class TransientError(DeliveryError):
    """429, 5xx, timeouts, connection resets: retry with backoff."""

    retryable = True

    def __init__(
        self, message: str, *, status: int | None = None, retry_after: float | None = None
    ) -> None:
        super().__init__(message, status=status)
        self.retry_after = retry_after


class PermanentError(DeliveryError):
    """Hard 4xx (bad payload, auth, not found): do not retry, let SQS redrive."""

    retryable = False


def classify_response(response: httpx.Response, policy: RetryPolicy) -> DeliveryError | None:
    """Return the error a response represents, or None when it succeeded."""
    if response.is_success:
        return None
    status = response.status_code
    body = response.text[:200]
    if status in policy.retry_on_status or status >= 500:
        retry_after = _parse_retry_after(response.headers.get("Retry-After"))
        return TransientError(f"HTTP {status}: {body}", status=status, retry_after=retry_after)
    return PermanentError(f"HTTP {status}: {body}", status=status)


def classify_exception(exc: Exception) -> DeliveryError:
    if isinstance(exc, DeliveryError):
        return exc
    if isinstance(exc, httpx.TimeoutException | httpx.NetworkError | httpx.RemoteProtocolError):
        return TransientError(f"{type(exc).__name__}: {exc}")
    return PermanentError(f"{type(exc).__name__}: {exc}")


def _parse_retry_after(value: str | None) -> float | None:
    if not value:
        return None
    try:
        return max(0.0, float(value))
    except ValueError:
        return None


def backoff_delay(attempt: int, policy: RetryPolicy, rng: random.Random | None = None) -> float:
    """Full-jitter delay for ``attempt`` (1-based): uniform(0, min(cap, base * mult^(n-1)))."""
    if attempt < 1:
        raise ValueError("attempt is 1-based")
    ceiling = min(policy.max_seconds, policy.base_seconds * (policy.multiplier ** (attempt - 1)))
    return (rng or random).uniform(0, ceiling)


def backoff_schedule(policy: RetryPolicy, rng: random.Random | None = None) -> Iterator[float]:
    for attempt in range(1, policy.max_attempts):
        yield backoff_delay(attempt, policy, rng)


@dataclass
class RetryOutcome:
    attempts: int
    delays: list[float] = field(default_factory=list)


def retry_call(
    fn: Callable[[], T],
    policy: RetryPolicy,
    *,
    sleep: Callable[[float], None] = time.sleep,
    rng: random.Random | None = None,
    on_retry: Callable[[int, float, TransientError], None] | None = None,
) -> tuple[T, RetryOutcome]:
    """Call ``fn`` until it succeeds, raises a PermanentError, or exhausts attempts.

    Only ``TransientError`` triggers a retry. When ``max_attempts`` is exhausted the last
    ``TransientError`` is re-raised so the caller (the SQS worker) can leave the message
    for redrive instead of acknowledging it.
    """
    outcome = RetryOutcome(attempts=0)
    while True:
        outcome.attempts += 1
        try:
            return fn(), outcome
        except Exception as exc:
            err = classify_exception(exc)
            if not err.retryable or outcome.attempts >= policy.max_attempts:
                raise err from exc
            assert isinstance(err, TransientError)
            delay = backoff_delay(outcome.attempts, policy, rng)
            if err.retry_after is not None:
                delay = min(policy.max_seconds, max(delay, err.retry_after))
            outcome.delays.append(delay)
            if on_retry:
                on_retry(outcome.attempts, delay, err)
            sleep(delay)
